lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -356,4 +398,56 @@ sub _update_scheduled_jobs {
     }
 }
 
+# serializes the sequence of directly chained jobs inside the specified $cluster_info starting from $first_job_id
+# remarks:
+#  * Direct dependency chains might be interrupted by regularily chained dependencies. Jobs not reachable from $first_job_id
+#    via directly chained dependencies nodes are not included.
+#  * Provides a 'flat' list of involved job IDs as 2nd return value.
+#  * See subtest 'serialize sequence of directly chained dependencies' in t/05-scheduler-dependencies.t for examples.
+sub _serialize_directly_chained_job_sequence {
+    my ($first_job_id, $cluster_info) = @_;
+
+    my %visited = ($first_job_id => 1);
+    my $sequence
+      = _serialize_directly_chained_job_sub_sequence([$first_job_id], \%visited,
+        $cluster_info->{$first_job_id}->{directly_chained_children},
+        $cluster_info);
+    return ($sequence, [keys %visited]);
+}
+sub _serialize_directly_chained_job_sub_sequence {
+    my ($output_array, $visited, $child_job_ids, $cluster_info) = @_;
+
+    for my $current_job_id (@$child_job_ids) {
+        die "cycle at $current_job_id" if $visited->{$current_job_id}++;
```

Comment:
Can you please make that die message a bit more explicit or informative? E.g. "Dependency cycle detected in directly-chained job sub sequence at job ID '$current_job_id'" ?

Reply:
This die is only supposed to fire if someone puts in a cycle into the database by manually modifying the SQL. I just want to prevent the worst case here (system freezes because the process allocates all available memory). The user facing validation already happens when inserting dependencies into the database when scheduling an ISO (test for this are present in this PR).

Reply:
To me the message sounds like it will try again - that's not what it does, though, I think?